x-pack/auditbeat/module/system/host/host.go

```diff
@@ -5,23 +5,116 @@
 package host
 
 import (
+	"bytes"
+	"encoding/gob"
+	"io"
 	"net"
+	"time"
 
+	"github.com/OneOfOne/xxhash"
+	"github.com/joeshaw/multierror"
 	"github.com/pkg/errors"
 
+	"github.com/elastic/beats/auditbeat/datastore"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/common/cfgwarn"
-	"github.com/elastic/beats/metricbeat/mb"
-
 	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
 	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
 )
 
 const (
 	moduleName    = "system"
 	metricsetName = "host"
+
+	bucketName        = "host.v1"
+	bucketKeyLastHost = "lastHost"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+
+	eventActionHost        = "host"
+	eventActionIdChanged   = "host_id_changed"
+	eventActionReboot      = "reboot"
+	eventActionHostChanged = "host_changed"
 )
 
+// Host represents information about a host.
+type Host struct {
+	info types.HostInfo
+	// Uptime() in types.HostInfo recalculates the uptime every time it is called -
+	// so storing it permanently here.
+	uptime time.Duration
+	addrs  []net.Addr
+	macs   []net.HardwareAddr
+}
+
+// changeDetectionHash creates a hash of selected parts of the host information.
+// This is used later to detect changes to a host over time.
+func (host *Host) changeDetectionHash() uint64 {
+	h := xxhash.New64()
+
+	mapstr := host.toMapStr()
+
+	// We detect changes to the ID separately, so ignore here.
+	mapstr.Delete("id")
+
+	// Changes in uptime and boottime indicate a system reboot,
+	// which we report separately.
+	mapstr.Delete("uptime")
+	mapstr.Delete("boottime")
+
+	h.WriteString(mapstr.String())
```

Comment:
The combination of creating a map and then calling `MapStr.String()` which IIRC marshals the data as JSON will be a fairly expensive operation. I think it would be better to directly write the values that should be included in the hash calculation into the `Hash` via its `io.Writer` interface.

Reply:
You're right, I've changed to using the fields directly.